csrc/preseg_passes/remove_bcast_squeeze.cpp

```diff
@@ -8,46 +8,296 @@
+    } else {
+      NVF_ERROR(
+          next[next_pos] == AxisOp::SQUEEZE &&
+          prev[prev_pos] == AxisOp::BROADCAST);
+      // Skip this op, since it is a "broadcast then squeeze" pattern
+    }
+    prev_pos++;
+    next_pos++;
+  }
+  return ops;
+}
 
-    // step-2: find and remove squeeze + broadcast pattern
-    // before: Y = squeeze(X); Z = broadcast(Y);  M = someOp(Z)
-    // after : M = someOp(X)
-    // conditions: (1) broadcast and squeeze have the same dim flags
-    if (auto bcast = dynamic_cast<BroadcastOp*>(expr)) {
-      if (auto squeeze = dynamic_cast<SqueezeOp*>(bcast->in()->definition())) {
-        if (bcast->getBroadcastDimFlags() == squeeze->getSqueezeDimFlags()) {
-          ir_utils::replaceValInAllExprInputsAndFusionOutputs(
-              bcast->out(), squeeze->in());
+TensorView* maybeDoReplacement(TensorView* orig) {
+  Expr* second = orig->definition();
+  if (!isReplaceableExpr(second)) {
+    return orig;
+  }
+  Expr* first = second->input(0)->definition();
+  if (!isReplaceableExpr(first)) {
+    return orig;
+  }
+
+  AxisOps first_ops = exprToAxisOps(first);
+  AxisOps second_ops = exprToAxisOps(second);
+  AxisOps simplified_ops = composeOps(first_ops, second_ops);
+  std::optional<AxisOp> simple_op_type_opt =
+      getSimplifiedOpType(simplified_ops);
+  if (!simple_op_type_opt.has_value()) {
+    return orig;
+  }
+  TensorView* replacement = orig;
+  if (simplified_ops == first_ops) {
+    // The second op was simply a "Set" operation, so we just skip it
+    replacement = first->output(0)->as<TensorView>();
+  } else {
+    TensorView* input_tv = first->input(0)->as<TensorView>();
+    switch (simple_op_type_opt.value()) {
+      case AxisOp::PRESERVE:
+        // This is equivalent to a set Op
+        replacement = input_tv;
+        // Check that parallelization is consistent for replacement. We do not
+        // want to alter the parallelization of input_tv here; we only
+        // parallelize _new_ TVs. If the parallelization is inconsistent, we
+        // simply refuse to make the replacement.
+        if (!hasCompatibleParallelization(orig, replacement)) {
+          return orig;
         }
-      }
+        break;
+      case AxisOp::SQUEEZE:
+        replacement = squeeze(input_tv, nonPreservedDims(simplified_ops));
+        break;
+      case AxisOp::BROADCAST:
+        replacement = broadcast(input_tv, nonPreservedDims(simplified_ops));
+        break;
+    }
+  }
+  NVF_ERROR(replacement != orig, "Expected non-trivial replacement");
+  ir_utils::replaceValInAllExprInputsAndFusionOutputs(orig, replacement);
+  std::vector<IterDomain*> old_loop =
+      TensorDomain::noReductions(orig->getLoopDomain());
+  std::vector<IterDomain*> new_loop =
+      TensorDomain::noReductions(replacement->getLoopDomain());
+  NVF_ERROR(new_loop.size() == old_loop.size());
+  for (size_t i : c10::irange(old_loop.size())) {
+    if (old_loop[i]->isParallelized()) {
```

Comment:
Here in pre-segment pass, device parallelize is the only parallel type we have. Can we just use parallel type device and add NVF to ensure no other unexpected parallel types?